### PR TITLE
Add "switch_map" operator

### PR DESCRIPTION
**Pull request by tcw165**

### CODE REVIEW

File: src/rpp/rpp/operators/fwd/switch_map.hpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+concept switch_map_callable = std::invocable<Fn, Type> && constraint::observable<std::invoke_result_t<Fn, Type>>;
+
+template<constraint::decayed_type Type, switch_map_callable<Type> Callable>
+auto switch_map_impl(auto&& observable, Callable&& callable);
+
+template<constraint::decayed_type Type, typename SpecificObservable>
+struct member_overload<Type, SpecificObservable, switch_map_tag>
+{
+    /**
+    * \brief Transform emissions to observables via provided function and then merge emissions from such an observables
```
Comment (victimsnino) [inline src/rpp/rpp/operators/fwd/switch_map.hpp:33]: Please, change description to actual one (you can take some part from `switch_on_next`)

File: src/rpp/rpp/operators/fwd/switch_map.hpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    *
+    * \warning According to observable contract (https://reactivex.io/documentation/contract.html) emissions from any observable should be serialized, so, resulting observable uses mutex to satisfy this requirement
```
Comment (victimsnino) [inline src/rpp/rpp/operators/fwd/switch_map.hpp:35]: This one is not actual for this operator

File: src/rpp/rpp/operators/fwd/switch_map.hpp [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    * 
+    * \marble switch_map
```
Comment (victimsnino) [inline src/rpp/rpp/operators/fwd/switch_map.hpp:37]: Can you create more meaningul example? You can do some experiments [there](https://rx-marbles-online.herokuapp.com/) and to see docs [there](https://bitbucket.org/achary/rx-marbles/src/762c9519e58f4cb914c9187c0b3019d5c665b933/docs/syntax.md)

File: src/rpp/rpp/operators/fwd/switch_map.hpp [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        {
+            source observable                   : +--1--2--3--|
+            operator "switch_map: x=>just(x,x+1)" : +--12-23-34-|
+        }
+    *
+    * \details Actually it makes `map` and then `merge`.
```
Comment (victimsnino) [inline src/rpp/rpp/operators/fwd/switch_map.hpp:43]: map + switch_on_next

File: src/rpp/rpp/operators/switch_map.hpp [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+//
+// Project home: https://github.com/victimsnino/ReactivePlusPlus
+//
+
+#pragma once
+
+#include <rpp/operators/fwd/switch_map.hpp>
+
+#include <rpp/operators/map.hpp>
+#include <rpp/operators/merge.hpp>
```
Comment (victimsnino) [inline src/rpp/rpp/operators/switch_map.hpp:16]: you don't need it

File: src/samples/doxygen/switch_map.cpp [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+#include <iostream>
+
+/**
+ * \example switch_map.cpp
+ **/
+int main()
+{
+    //! [switch_map]
+    rpp::source::just(1, 2, 3)
+            .switch_map([](int val) { return rpp::source::from_iterable(std::vector(val, val)); })
```
Comment (victimsnino) [inline src/samples/doxygen/switch_map.cpp:11]: can you add something with `never` to show this important moment?

File: src/tests/test_switch_map.cpp [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+#include <rpp/observables/dynamic_observable.hpp>
+#include <rpp/operators/switch_map.hpp>
+#include <rpp/sources/just.hpp>
+#include <rpp/sources/error.hpp>
+#include <rpp/sources/empty.hpp>
+#include <rpp/sources/never.hpp>
+
+#include "mock_observer.hpp"
+
+SCENARIO("switch_map acts like flat_map except it unsubscribes the previous source when new source emits", "[flat_map]")
```
Comment (victimsnino) [inline src/tests/test_switch_map.cpp:21]:
```suggestion
SCENARIO("switch_map acts like flat_map except it unsubscribes the previous source when new source emits", "[switch_map]")
```

File: src/tests/test_switch_map.cpp [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+                .subscribe(mock);
+            THEN("subscriber obtains values from observables obtained via switch_map")
+            {
+                CHECK(mock.get_total_on_next_count() == 0);
+                CHECK(mock.get_on_completed_count() == 0);
+                CHECK(mock.get_on_error_count() == 0);
+                CHECK(sub.is_subscribed());
+            }
+        }
+        WHEN("subscribe on it via switch_map with mix of never + observable")
```
Comment (victimsnino) [inline src/tests/test_switch_map.cpp:72]: can't see any difference with previous "WHEN"

File: src/tests/test_switch_map.cpp [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+                    .subscribe(mock);
+            THEN("subscriber obtains values from observables obtained via switch_map")
+            {
+                CHECK(mock.get_total_on_next_count() == 0);
+                CHECK(mock.get_on_completed_count() == 0);
+                CHECK(mock.get_on_error_count() == 0);
+                CHECK(sub.is_subscribed());
+            }
+        }
+        WHEN("subscribe on it via switch_map with never + observables")
```
Comment (victimsnino) [inline src/tests/test_switch_map.cpp:84]: Can you add same test, but when never is last and no `on_completed` call as a result?

**Review by victimsnino — CHANGES_REQUESTED** [review]

File: src/rpp/rpp/operators/fwd/switch_map.hpp (hunk 3 above)
Comment (tcw165) [reply src/rpp/rpp/operators/fwd/switch_map.hpp:37]: Yup, I will.

File: src/rpp/rpp/operators/fwd/switch_map.hpp (hunk 4 above)
Comment (tcw165) [reply src/rpp/rpp/operators/fwd/switch_map.hpp:43]: Thanks for eye-picking the old doc in the comment. :)

File: src/rpp/rpp/operators/switch_map.hpp (hunk 5 above)
Comment (tcw165) [reply src/rpp/rpp/operators/switch_map.hpp:16]: I'll remove them :)

File: src/samples/doxygen/switch_map.cpp (hunk 6 above)
Comment (tcw165) [reply src/samples/doxygen/switch_map.cpp:11]: Sounds good!

File: src/tests/test_switch_map.cpp (hunk 9 above)
Comment (tcw165) [reply src/tests/test_switch_map.cpp:73]: That's a very good point. I'll add it.

File: src/tests/test_switch_map.cpp [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+            obs.switch_map([&](int i) {
+                    if (i == 1) {
+                        return rpp::source::never<int>()
+                            // "lift" is used for capturing the unsubscribe event
+                            .lift([&](const auto& sub) {
+                                sub.get_subscription().add([&]() {
+                                    ++count_of_unsubscribed;
+                                });
+                                return sub;
+                            })
```
Comment (tcw165) [inline src/tests/test_switch_map.cpp:86]: @victimsnino Hey, can you help me understand why my compiler complains the "lift" is not found and so there's no matching candidate function?

File: src/tests/test_switch_map.cpp [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+                CHECK(mock.get_on_error_count() == 0);
+            }
+            THEN("subscription is still subscribed since no inner observable completes")
+            {
+                CHECK(sub.is_subscribed());
+            }
+        }
+        WHEN("subscribe on it via switch_map with [never, just...]")
+        {
+            std::atomic_size_t count_of_unsubscribed{0};
```
Comment (victimsnino) [inline src/tests/test_switch_map.cpp:79]: Not sure if it should be atomic due to you don't have any threading there =)

**Review by victimsnino — APPROVED** [review]

**Pull request merged**